raimitigations/dataprocessing/imputer/basic_imputer.py

```diff
@@ -231,6 +231,21 @@ def _transform(self, df: pd.DataFrame):
 
         :param df: the full dataset being transformed.
         """
+        if self.none_status is True:
+            col_nan_status = df.isna().any()
+            col_with_nan = []
+            for i, value in enumerate(col_nan_status.index):
+                if col_nan_status[value]:
+                    if type(value) == int:
+                        col_with_nan.append(i)
+                    else:
+                        col_with_nan.append(value)
+            self.print_message(
+                f"No columns specified for imputation. These columns "
+                + f"have been automatically identified at transform time:\n{col_with_nan}"
+            )
+            self.col_impute = col_with_nan
+
```

Comment:
I see your point, but a few reasons why I think we should maintain a flexibility in detecting missing values automatically both at fit and transform:

- All of these classes are heavily based on sklearn's modules and we should try to maintain as much as possible of the functionality and flexibility present in sklearn's APIs. Sklearn's imputers automatically look for all missing values to impute and they do in fact allow for datasets to have different columns with missing values at fit and transform.
- Both [SimpleImputer ](https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html)and [IterativeImputer ](https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html)actually highlight this in the main example in their documentations, both examples show how column 0 wasn't missing any values at fit, but will be imputed at transform:
```
import numpy as np
from sklearn.impute import SimpleImputer
imp_mean = SimpleImputer(missing_values=np.nan, strategy='mean')
imp_mean.fit([[7, 2, 3], [4, np.nan, 6], [10, 5, 9]])

X = [[np.nan, 2, 3], [4, np.nan, 6], [10, np.nan, 9]]
print(imp_mean.transform(X))

output:
[[ 7.   2.   3. ]
 [ 4.   3.5  6. ]
 [10.   3.5  9. ]]
```
```
import numpy as np
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
imp_mean = IterativeImputer(random_state=0)
imp_mean.fit([[7, 2, 3], [4, np.nan, 6], [10, 5, 9]])

X = [[np.nan, 2, 3], [4, np.nan, 6], [10, np.nan, 9]]
imp_mean.transform(X)

output:
array([[ 6.9584...,  2.       ,  3.        ],
       [ 4.       ,  2.6000...,  6.        ],
       [10.       ,  4.9999...,  9.        ]])
```

- I believe the point here is to maintain that flexibility and choice with the user, we provide them with clear messages but it should be up to them, however it affects their results as they are more familiar with their data. As long as they maintain the same data structure (column names, dataset length..etc) for fit and transform datasets, missing values should be imputed wherever they are at transform. 
- Note that this is specifically for when we adopt the default behavior of sklearn's apis, when users choose `col_impute=None` and we automatically detect columns with missing data.

Reply:
In this case, then always set ``self.col_impute`` as the list of all columns. This way, if users don't specify anything, we'll allow all columns to be imputed during transform(), even for columns that had no missing values during fit().

Actually, it seems like the ``self.col_impute`` attribute isn't important at all to the Iterative and KNN imputers. IterativeImputer uses the ``skip_complete`` parameter for that, while the KNNImputer uses all available columns either way. So, I think you should consider making the self.col_impute parameter something specific to BasicImputer, and not use this for the other 2 imputers.

Finally, the way I created the BasicImputer, since I allow users to use different imputation strategies for each column, I end up creating one ``SimpleImputer`` for each column. This way, if no columns are specified, I'll create one ``SimpleImputer`` for each column with missing value during the fit() call. If the user then pass a new dataset to the transform() method with missing values in a new column, it will give an error using your strategy, since I didn't create a ``SimpleImputer`` for that column. For example, the following code will give an error:

```
import numpy as np
import pandas as pd
from raimitigations.dataprocessing import BasicImputer

df1 = pd.DataFrame({'A':[3, np.nan, 7, 10],
                    'B':[1, 2, 3, 4],
                    'C':[60, 40, 70, 50],
                    'D':[100, 99, np.nan, 110]
                })
imp = BasicImputer()
imp.fit(df1)
new_df1 = imp.transform(df1)

df2 = pd.DataFrame({'A':[6, 4, 2, np.nan],
                    'B':[0, 3, 2, 1],
                    'C':[50, np.nan, 80, 45],
                    'D':[104, 106, 95, 98]
                })
new_df2 = imp.transform(df2)
print(new_df2)
```

Reply:
Okay so I've dug a bit deeper into the behavior of each imputer when new columns with missing values appear at transform not present at fit time and here's what I found/concluded:

**Iterative Imputer**:
- `skip_complete` was added by sklearn to basically avoid the situation of fitting an estimator at fit time and transforming on different data, which obviously can be buggy. So, using `skip_complete=False` (the default) will refit an estimator and impute these new columns correctly. Now, if the user wanted faster performance, then they would use `skip_complete=True` and new missing values would be imputed using the mean of the column at fit. Both of these are suitable non-buggy behavior, leave the flexibility of choice to the user and don't contradict, but work well with the current setup of our new class while maintaining consistency with sklearn's behavior.
- I'm going to leave `col_impute` as a parameter, not only to be consistent with the parent class, but also because its useful for other scenarios, for e.g.: if users would like to exclude certain columns from imputation (especially categorical ones).
- Not recalculating missing columns at transform and using `col_impute=all_columns` would have the same behavior in this case, however, recalculating columns with missing values is useful as part of the transform function in other places of the code, such as separating the encoding reverse transform of encoded imputed values and non-imputed categorical values, in the case of errors in the reverse transformation of imputed values.

Reply:
Okay so I've dug a bit deeper into the behavior of each imputer when new columns with missing values appear at transform not present at fit time and here's what I found/concluded:

**Iterative Imputer**:
- `skip_complete` [was added by sklearn](https://github.com/scikit-learn/scikit-learn/pull/14806) to basically avoid the situation of fitting an estimator at fit time and transforming on different data, which obviously can be buggy. So, using `skip_complete=False` (the default) will refit an estimator and impute these new columns correctly. Now, if the user wanted faster performance, then they would use `skip_complete=True` and new missing values would be imputed using the mean of the column at fit. Both of these are suitable non-buggy behaviors, leave the flexibility of choice to the user and don't contradict, but work well with the current setup of our new class while maintaining consistency with sklearn's behavior.
- I'm going to leave `col_impute` as a parameter, not only to be consistent with the parent class, but also because its useful for other scenarios, for e.g.: if users would like to exclude certain columns from imputation (especially categorical ones).
- Not recalculating missing columns at transform and using `col_impute=all_columns` would have the same behavior in this case, however, recalculating columns with missing values is useful as part of the transform function in other places of the code, such as separating the encoding reverse transform of encoded imputed values and non-imputed categorical values, in the case of errors in the reverse transformation of imputed values.

**KNN Imputer**:
- If new missing values of features arise, sklearn's API will still use the same KNN procedure, using the mean of the k-nearest neighbors from the fit time, so the behavior isn't buggy nor does it change.
- I will keep `col_impute` as a parameter for the same reasons above.
- Recalculating missing values at transform or simply setting `col_impute=all_columns`, again results in the same behavior, but recalculating again has the advantage as its useful in other parts of the code. 

**Simple Imputer**:
- sklearn's API can in fact handle new missing value features and performs the same behavior as it would on features with missing values at fit. 
-  I think the setup of creating a new imputer for each column isn't very efficient especially in this case, but I see why it's setup as such given the param dicts. 
- You're right recalculating will error for the BasicImputer, in order to match sklearn's api, we'd have to do an exception handling followed by another fit and transform, but that isn't very straightforward given the param dicts for different types of data. Although I can add this behavior in a function at transform (let me know if you think it's the better option), instead, to minimize major changes, I'm going to simply remove the new recalculation of col_impute at transform for the BasicImputer maintaining the previous behavior.